File: Java/commons/repository-interfaces/src/main/java/org/cqrs101/RepositoryHelperFactory.java
package org.cqrs101;

import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class RepositoryHelperFactory {

    private final RepositoryHelper repositoryHelper;
    private final ConcurrentHashMap<Class, RepositoryHelper> repositoryHelperInstances;

    public RepositoryHelperFactory(RepositoryHelper repositoryHelper) {
        this.repositoryHelper = repositoryHelper;
        repositoryHelperInstances = new ConcurrentHashMap<>();
    }

    public RepositoryHelper getHelper(Class clazz) {
        if (!repositoryHelperInstances.containsKey(clazz)) {
            repositoryHelperInstances.putIfAbsent(clazz, repositoryHelper.create(clazz));
        }
        return repositoryHelperInstances.get(clazz);
    }
}

File: Net/Commons/Repository.Interfaces/RepositoryHelperFactory.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Text;
using Utils;

namespace Repository.Interfaces
{
    public class IRepositoryHelperFactory : ISingleton
    {
        private IRepositoryHelper repositoryHelper;
        private ConcurrentDictionary<Type, IRepositoryHelper> repositoryHelperInstances;

        public IRepositoryHelperFactory(IRepositoryHelper repositoryHelper)
        {
            this.repositoryHelper = repositoryHelper;
            repositoryHelperInstances = new ConcurrentDictionary<Type, IRepositoryHelper>();
        }

        public IRepositoryHelper getHelper(Type clazz)
        {
            if (!repositoryHelperInstances.ContainsKey(clazz))
            {
                repositoryHelperInstances[clazz] = repositoryHelper.Create(clazz);
            }
            return repositoryHelperInstances[clazz];
        }
    }
}
